README.md
# auditd

auditd - The Linux Audit daemon

## Requirements

### Platform

* Debian (= 8.1)
* CentOS (= 7.1)
* Ubuntu (= 14.04)

## Recipes

### auditd::default

Installs the auditd package and starts, enables the auditd service.

## Resource/Provider

### `auditd_config`

This resource manages the `auditd` service configuration file. All configuration parameters are supported. If you find one is missing please file an issue.

#### Actions

- `:write` _(default)_

#### Properties

<table>
 <thead>
    <tr>
      <td>Name</td>
      <td>Description</td>
      <td>Default</td>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><code>:log_file</code></td>
      <td>Full path to the log file where audit records will be stored</td>
      <td><code>/var/log/audit/audit.log</code></td>
    </tr>
    <tr>
      <td><code>:log_format</code></td>
      <td>Describes how the information should be stored on disk</td>
      <td><code>raw</code></td>
    </tr>
    <tr>
      <td><code>:log_group</code></td>
      <td>Specifies the group that is applied to the log file's permissions</td>
      <td><code>root</code></td>
    </tr>
    <tr>
      <td><code>:priority_boost</code></td>
      <td>Tells the audit daemon how much of a priority boost it should take</td>
      <td><code>4</code></td>
    </tr>
    <tr>
      <td><code>:flush</code></td>
      <td>Valid values are none, incremental, data, and sync</td>
      <td><code>incremental</code></td>
    </tr>
    <tr>
      <td><code>:freq</code></td>
      <td>Tells the audit daemon how many records to write before issuing an explicit flush to disk command</td>
      <td><code>20</code></td>
    </tr>
    <tr>
      <td><code>:num_logs</code></td>
      <td>The number of log files to keep</td>
      <td><code>5</code></td>
    </tr>
    <tr>
      <td><code>:disp_qos</code></td>
      <td>Controls whether blocking/lossless or non-blocking/lossy communication between the audit daemon and the dispatcher</td>
      <td><code>lossy</code></td>
    </tr>
    <tr>
      <td><code>:dispatcher</code></td>
      <td>Dispatcher program started by the audit daemon</td>
      <td><code>/sbin/audispd</code></td>
    </tr>
    <tr>
      <td><code>:name_format</code></td>
      <td>Controls how computer node names are inserted into the audit event  stream</td>
      <td><code>none</code></td>
    </tr>
    <tr>
      <td><code>:name_value</code></td>
      <td>String that identifies the machine</td>
      <td><code>nil</code></td>
    </tr>
    <tr>
      <td><code>:max_log_file</code></td>
      <td>Maximum file size in megabytes</td>
      <td><code>6</code></td>
    </tr>
    <tr>
      <td><code>:max_log_file_action</code></td>
      <td>Action to take when the max file size limit has been reached</td>
      <td><code>rotate</code></td>
    </tr>
    <tr>
      <td><code>:action_mail_acct</code></td>
      <td>A valid email address or alias</td>
      <td><code>root</code></td>
    </tr>
    <tr>
      <td><code>:space_left</code></td>
      <td>Threshold for triggering the <code>space_left_action</code></td>
      <td><code>75</code></td>
    </tr>
    <tr>
      <td><code>:space_left_action</code></td>
      <td>Action to take if the system is low on disk space</td>
      <td><code>syslog</code></td>
    </tr>
    <tr>
      <td><code>:admin_space_left</code></td>
      <td>Threshold for triggering the <code>admin_space_left_action</code></td>
      <td><code>50</code></td>
    </tr>
    <tr>
      <td><code>:admin_space_left_action</code></td>
      <td>Action to take if the system is low on disk space</td>
      <td><code>suspend</code></td>
    </tr>
    <tr>
      <td><code>:disk_full_action</code></td>
      <td>Action to take if partition auditd is writing to has become full</td>
      <td><code>suspend</code></td>
    </tr>
    <tr>
      <td><code>:disk_error_action</code></td>
      <td>Action to take if an error is detected writing to disk</td>
      <td><code>suspend</code></td>
    </tr>
    <tr>
      <td><code>:tcp_listen_port</code></td>
      <td>Causes auditd to listen on the corresponding TCP port</td>
      <td><code>nil</code></td>
    </tr>
    <tr>
      <td><code>:tcp_listen_queue</code></td>
      <td>Indicates how many pending connections are allowed</td>
      <td><code>5</code></td>
    </tr>
    <tr>
      <td><code>:tcp_max_per_addr</code></td>
      <td>Indicates how many concurrent connections from one IP address is allowed</td>
      <td><code>1</code></td>
    </tr>
    <tr>
      <td><code>:use_libwrap</code></td>
      <td>Determines whether or not to use tcp_wrappers</td>
      <td><code>yes</code></td>
    </tr>
    <tr>
      <td><code>:tcp_client_ports</code></td>
      <td>Indicates which client ports are allowed</td>
      <td><code>nil</code></td>
    </tr>
    <tr>
      <td><code>:tcp_client_max_idle</code></td>
      <td>Indicates the number of seconds that a client may be idle</td>
      <td><code>0</code></td>
    </tr>
    <tr>
      <td><code>:enable_krb5</code></td>
      <td>Determines if Kerberos 5 will be used for authentication and encryption</td>
      <td><code>no</code></td>
    </tr>
    <tr>
      <td><code>:krb5_principal</code></td>
      <td>The principal for this server</td>
      <td><code>auditd</code></td>
    </tr>
    <tr>
      <td><code>:krb5_key_file</code></td>
      <td>Location of the key for this client's principal</td>
      <td><code>nil</code></td>
    </tr>
  </tbody>
</table>

### `auditd_ruleset`

This resource is used to define cumulative rulesets across a `chef-client` run.

#### Actions

- `:include` _(default)_

#### Properties

<table>
  <tr>
    <td>Attribute</td>
    <td>Description</td>
    <td>Default</td>
  </tr>
  <tr>
    <td><code>:comment</code></td>
    <td>Descriptive comment to associate with the rules</td>
    <td><code>nil</code></td>
  </tr>
  <tr>
    <td><code>:rules</code></td>
    <td>An <code>Array</code> of <code>String</code> objects representing rules to add</td>
    <td><code>[]</code></td>
  </tr>
</table>

### `auditd_rules`

This resource writes the collective rulesets defined across all `auditd_ruleset` resources to the specified rules file. If the rules have been updated, this resource will trigger a `reload` of the auditd service.

#### Actions

- `:include` _(default)_

## Examples

```ruby
# build the default rules
auditd_ruleset 'default' do
  comment 'First rule - delete all'
  rules ['-D']
end

auditd_ruleset 'buffers' do
  comment 'Increase the buffers to survive stress events.'
  rules ['-b 320']
end

# write the rules we defined above
auditd_rules '/etc/audit/audit.rules'
```


CHANGELOG.md
# auditd

## v0.1.1

* Add chefspec coverage for both rhel and debian platform families
* Send `:reload` notification to auditd when configuration is updated

## v0.1.0

* Initial Release


spec/spec_helper.rb
require 'chefspec'
require 'chefspec/berkshelf'

##################
# Monkey Patches #
##################

# silence all logging from chef-client
Chef::Log.init('/dev/null')

# silence 'warning: already initialized constant' stuff
$VERBOSE = nil

# silence the deprecation warnings
module Kernel
  def deprecated(*)
    # noop
  end
end


spec/unit/recipes/default_spec.rb
#
# Cookbook Name:: auditd
# Spec:: default
#
# Copyright (c) 2015 The Authors, All Rights Reserved.

require 'spec_helper'

describe 'auditd::default' do
  context 'When all attributes are default, on a debian-family system' do
    let(:chef_run) do
      ChefSpec::ServerRunner.new do |node, _|
        node.automatic['platform_family'] = 'debian'
      end.converge(described_recipe)
    end

    it 'installs auditd' do
      expect(chef_run).to install_package('auditd')
    end

    it 'starts and enables the auditd service' do
      expect(chef_run).to start_service('auditd')
      expect(chef_run).to enable_service('auditd')
    end
  end

  context 'When all attributes are default on a rhel-family system' do
    let(:chef_run) do
      ChefSpec::ServerRunner.new do |node, _|
        node.automatic['platform_family'] = 'rhel'
      end.converge(described_recipe)
    end

    it 'installs auditd' do
      expect(chef_run).to install_package('audit')
    end

    it 'starts and enables the auditd service' do
      expect(chef_run).to start_service('auditd')
      expect(chef_run).to enable_service('auditd')
    end
  end
end


test/fixtures/cookbooks/auditd_example/metadata.rb
name 'auditd_example'

depends 'auditd'


resources/ruleset.rb
resource_name :auditd_ruleset

property :rules, Array, default: []
property :comment, String, default: nil

default_action :include

action :include do
  check_run_state
  Chef::Log.debug "before: #{rulesets}"
  include_rules
  Chef::Log.debug "after: #{rulesets}"
  if has_rules?
    new_resource.updated_by_last_action(false)
  else
    new_resource.updated_by_last_action(true)
  end
end

def include_rules
  rulesets[name] = Mash.new({
    comment: comment,
    rules: rules
  })
end

def check_run_state
  node.run_state['auditd'] ||= {}
  node.run_state['auditd']['rulesets'] ||= {}
end

def rulesets
  node.run_state['auditd']['rulesets']
end

def has_rules?
  rules.each do |rule|
    return false unless existing_rule?(rule)
  end
  true
end

def current_rules
  ::File.open('/etc/audit/audit.rules').readlines.map(&:strip)
end

def existing_rule?(text)
  current_rules.include? text
end


resources/config.rb
resource_name :auditd_config

property :path,
  String,
  name_attribute: true

# This keyword specifies the full path name to the log file where audit records
# will be stored. It must be a regular file.
property :log_file,
  String,
  default: '/var/log/audit/audit.log'

# The log format describes how the information should be stored on disk. There
# are 2 options: raw and nolog.
property :log_format,
  String,
  default: 'RAW',
  regex: /raw|nolog/i

# This keyword specifies the group that is applied to the log file's
# permissions. The default is root. The group name can be either numeric or
# spelled out.
property :log_group,
  String,
  default: 'root'

# This is a non-negative number that tells the audit damon how much of a
# priority boost it should take. The default is 3. No change is 0.
property :priority_boost,
  Numeric,
  default: 4

# Valid values are none, incremental, data, and sync.
property :flush,
  String,
  default: 'incremental',
  regex: /none|incremental|data|sync/i

# This is a non-negative number that tells the audit damon how many records to
# write before issuing an explicit flush to disk command. this value is only
# valid when the flush keyword is set to incremental.
property :freq,
  Numeric,
  default: 20

# This keyword specifies the number of log files to keep if rotate is given as
# the max_log_file_action. If the number is less than 2, logs are not rotated.
# This number must be 99 or less. The default is 0 - which means no rotation.
property :num_logs,
  Numeric,
  default: 0,
  callbacks: { "check" => proc {|v| v.between?(0, 99) } }

# The dispatcher is a program that is started by the audit daemon when it starts
# up. It will pass a copy of all audit events to that application's stdin. Make
# sure you trust the application that you add to this line since it runs with
# root privileges.
property :dispatcher,
  String,
  default: '/sbin/audispd'

# This  option  controls  how  computer node names are inserted into the audit
# event stream. It has the following choices: none, hostname, fqd, numeric, and
# user.
property :name_format,
  String,
  default: 'none',
  regex: /none|hostname|fqd|numeric|user/

# @note this config value collides with the reserved `name` method
# This is the admin defined string that identifies the machine if user is given
# as the name_format option.
property :name_value,
  String,
  default: nil

# This option controls whether you want blocking/lossless or non-blocking/lossy
# communication between the audit daemon and the dispatcher. Valid values are:
# lossy and lossless. Lossy is the default value.
property :disp_qos,
  String,
  default: 'lossy',
  regex: /loss(y|less)/i

# This keyword specifies the maximum file size in megabytes. When this limit is
# reached, it will trigger a configurable action. The value given must be
# numeric.
property :max_log_file,
  Numeric,
  default: 6

# This parameter tells the system what action to take when the system has
# detected that the max file size limit has been reached. Valid values are
# ignore, syslog, suspend, rotate and keep_logs. If set to ignore, the audit
# daemon does nothing.
property :max_log_file_action,
  String,
  default: 'rotate',
  regex: /ignore|syslog|suspend|rotate|keep_logs/i

# This option should contain a valid email address or alias. The default address
# is root. If the email address is not local to the machine, you must make sure
# you have email properly configured on your machine and network. Also, this
# option requires that /usr/lib/sendmail exists on the machine.
property :action_mail_acct,
  String,
  default: 'root'

# This is a numeric value in megabytes that tells the audit daemon when to
# perform a configurable action because the system is starting to run low on
# disk space.
property :space_left,
  Numeric,
  default: 75

# This parameter tells the system what action to take when the system has
# detected that it is starting to get low on disk space. Valid values are:
# ignore, syslog, email, suspend, single, and halt.
property :space_left_action,
  String,
  default: 'syslog',
  regex: /ignore|syslog|email|suspend|single|halt/i

# This is a numeric value in megabytes that tells the audit daemon when to
# perform a configurable action because the system is running low on disk space.
property :admin_space_left,
  Numeric,
  default: 50

# This parameter tells the system what action to take when the system has
# detected that it is low on disk space. Valid values are ignore, syslog, email,
# suspend, single, and halt.
property :admin_space_left_action,
  String,
  default: 'suspend',
  regex: /ignore|syslog|email|suspend|single|halt/i

# This parameter tells the system what action to take when the system has
# detected that the partition to which log files are written has become full.
# Valid values are ignore, syslog, suspend, single, and halt.
property :disk_full_action,
  String,
  default: 'suspend',
  regex: /ignore|syslog|suspend|single|halt/i

# This parameter tells the system what action to take whenever there is an error
# detected when writing audit events to disk or rotating logs. Valid values are
# ignore, syslog, suspend, single, and halt.
property :disk_error_action,
  String,
  default: 'suspend',
  regex: /ignore|syslog|suspend|single|halt/i

# This is a numeric value in the range 1..65535 which, if specified, causes
# auditd to listen on the corresponding TCP port for audit records from remote
# systems.
property :tcp_listen_port,
  Numeric,
  default: nil,
  callbacks: { "check" => proc {|v| v.between?(1, 65535) } }

# This  is  a  numeric  value  which indicates how many pending (requested but
# unaccepted) connections are allowed. The default is 5.
property :tcp_listen_queue,
  Numeric,
  default: 5

# This is a numeric value which indicates how many concurrent connections from
# one IP address is allowed.  The default is 1 and the maximum is 1024.
property :tcp_max_per_addr,
  Numeric,
  default: 1,
  callbacks: { "check" => proc {|v| v.between?(1, 1024) } }

# This setting determines whether or not to use tcp_wrappers to discern
# connection attempts that are from allowed machines. Legal values are either
# yes, or no. The default value is yes.
property :use_libwrap,
  String,
  regex: /yes|no/i

# This parameter may be a single numeric value or two values separated by a dash
# (no spaces allowed). It indicates which client ports are allowed for incoming
# connections. If not specified, any port is allowed. Allowed values are
# 1..65535.
property :tcp_client_ports,
  String,
  regex: /\d+-?\d*/

# This parameter indicates the number of seconds that a client may be idle
# (i.e. no data from them at all) before auditd complains. This is used to close
# inactive connections if the client machine has a problem where it cannot
# shutdown the connection cleanly. Note that this is a global setting, and must
# be higher than any individual client heartbeat_timeout setting, preferably by
# a factor of two.  The default is zero, which disables this check.
property :tcp_client_max_idle,
  Numeric,
  default: 0

# If set to "yes", Kerberos 5 will be used for authentication and encryption.
# The default is "no".
property :enable_krb5,
  String,
  default: 'no',
  regex: /yes|no/i

# This is the principal for this server. The default is "auditd". Given this
# default, the server will look for a key named like auditd/hostname@EXAMPLE.COM
# stored in /etc/audit/audit.key to authenticate itself, where hostname is the
# canonical name for the server's host, as returned by a DNS lookup of its IP
# address.
property :krb5_principal,
  String,
  default: nil

# Location of the key for this client's principal. Note that the key file must
# be owned by root and mode 0400. The default is /etc/audit/audit.key
property :krb5_key_file,
  String,
  default: nil

default_action :write

action :write do
  template path do
    source 'auditd.conf.erb'
    cookbook 'auditd'
    owner 'root'
    group 'root'
    mode '0640'
    variables(
      config: self.dup
    )
  end
end
